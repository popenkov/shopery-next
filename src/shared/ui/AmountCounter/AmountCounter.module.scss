@import '@/shared/styles/_breakpoints';
@import '@/shared/styles/mixins/flexbox';
@import '@/shared/styles/mixins/box';
@import '@/shared/styles/mixins/button-reset';
@import '@/shared/styles/mixins/container';
@import '@/shared/styles/mixins/text-styles';


.AmountCounter {
  @include flex-center-vertical;

  gap: 2px;
  padding: 8px;
  border: 1px solid var(--monochrome-100);
  border-radius: 170px;


}

  .AmountCounterButtonIcon {
    @include box(14px);

    color: var(--monochrome-900);
  }

  .AmountCounterButton {
    @include button-reset("flex");
    @include box(34px);
    @include flex-center;

    background-color: var(--monochrome-50);
    border-radius: 50%;

    @media (hover: hover) {
      &:hover {
        background-color: var(--primary);

        .AmountCounterButtonIcon{
          color: var(--front-primary-inversed);
        }
      }
    }
  }

  .AmountCounterInputWrapper {
    @include flex-center-vertical;
  }

  .AmountCounterInput {
    @include body-m;

    width: 36px;
    min-width: 36px;
    text-align: center;
    border: none;

    &:focus {
      outline: none;
    }
  }