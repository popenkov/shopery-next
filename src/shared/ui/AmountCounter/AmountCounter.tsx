import { FC, useState } from 'react';
import cn from 'classnames';

import { MinusIcon, PlusIcon } from '@/shared/ui/icons';

import cls from './AmountCounter.module.scss';

type Props = {
  value: number;
  onChange: (value: number) => void;
  className?: string;
};

export const AmountCounter: FC<Props> = ({ value, onChange, className }) => {
  const [currentValue, setCurrentValue] = useState<number>(value || 1);

  const handleMinusButtonClick = () => {
    if (currentValue > 1) {
      setCurrentValue(currentValue - 1);
      onChange(currentValue - 1);
    }
  };

  const handlePlusButtonClick = () => {
    setCurrentValue(currentValue + 1);
    onChange(currentValue + 1);
  };

  return (
    <div className={cn(cls.AmountCounter, className)}>
      <button
        className={cls.AmountCounterButton}
        onClick={handleMinusButtonClick}
        data-testid="amount-counter-minus-button"
      >
        <MinusIcon className={cls.AmountCounterButtonIcon} />
      </button>
      <div className={cls.AmountCounterInputWrapper}>
        <input
          className={cls.AmountCounterInput}
          type="number"
          value={currentValue}
          min="1"
          readOnly
          data-testid="amount-counter-input"
        />
      </div>
      <button
        className={cls.AmountCounterButton}
        onClick={handlePlusButtonClick}
        data-testid="amount-counter-plus-button"
      >
        <PlusIcon className={cls.AmountCounterButtonIcon} />
      </button>
    </div>
  );
};
