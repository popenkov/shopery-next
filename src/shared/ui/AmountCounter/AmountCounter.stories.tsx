import type { Meta, StoryObj } from '@storybook/react';
import { fn } from '@storybook/test';
import { AmountCounter } from './AmountCounter';

const meta = {
  title: 'shared/AmountCounter',
  component: AmountCounter,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  args: { value: 5 },
  argTypes: {},
} satisfies Meta<typeof AmountCounter>;

export default meta;
type TStory = StoryObj<typeof meta>;

export const AmountCounterComponent: TStory = {
  decorators: [
    (Story) => (
      <div>
        <Story />
      </div>
    ),
  ],
  args: {
    value: 5,
    onChange: (value) => console.log('onChange:', value),
  },
};
