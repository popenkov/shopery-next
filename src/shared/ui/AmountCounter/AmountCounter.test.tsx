import React from 'react';
import { render, fireEvent } from '@testing-library/react';

import { AmountCounter } from './AmountCounter';

describe('AmountCounter component', () => {
  it('renders correctly', () => {
    const { container } = render(<AmountCounter value={1} onChange={() => {}} />);
    expect(container).toMatchSnapshot();
  });

  it('displays the initial value', () => {
    const value = 5;
    const { getByTestId } = render(<AmountCounter value={value} onChange={() => {}} />);
    const input = getByTestId('amount-counter-input');
    expect(input).toHaveValue(value);
  });

  it('calls onChange when the plus button is clicked', () => {
    const onChange = jest.fn();
    const { getByTestId } = render(<AmountCounter value={1} onChange={onChange} />);
    const plusButton = getByTestId('amount-counter-plus-button');
    fireEvent.click(plusButton);
    expect(onChange).toHaveBeenCalledTimes(1);
    expect(onChange).toHaveBeenCalledWith(2);
  });

  it('calls onChange when the minus button is clicked', () => {
    const onChange = jest.fn();
    const { getByTestId } = render(<AmountCounter value={2} onChange={onChange} />);
    const minusButton = getByTestId('amount-counter-minus-button');
    fireEvent.click(minusButton);
    expect(onChange).toHaveBeenCalledTimes(1);
    expect(onChange).toHaveBeenCalledWith(1);
  });

  it('does not call onChange when the minus button is clicked and the value is 1', () => {
    const onChange = jest.fn();
    const { getByTestId } = render(<AmountCounter value={1} onChange={onChange} />);
    const minusButton = getByTestId('amount-counter-minus-button');
    fireEvent.click(minusButton);
    expect(onChange).not.toHaveBeenCalled();
  });

  it('updates the input value when the plus button is clicked', () => {
    const { getByTestId } = render(<AmountCounter value={1} onChange={() => {}} />);
    const plusButton = getByTestId('amount-counter-plus-button');
    const input = getByTestId('amount-counter-input');
    fireEvent.click(plusButton);
    expect(input).toHaveValue(2);
  });

  it('updates the input value when the minus button is clicked', () => {
    const { getByTestId } = render(<AmountCounter value={2} onChange={() => {}} />);
    const minusButton = getByTestId('amount-counter-minus-button');
    const input = getByTestId('amount-counter-input');
    fireEvent.click(minusButton);
    expect(input).toHaveValue(1);
  });
});
