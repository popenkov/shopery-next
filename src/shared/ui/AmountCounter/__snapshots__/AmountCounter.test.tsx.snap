// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AmountCounter component renders correctly 1`] = `
<div>
  <div
    class="AmountCounter"
  >
    <button
      class="AmountCounterButton"
      data-testid="AmountCounter.minusButton"
    >
      <svg
        class="AmountCounterButtonIcon"
        fill="none"
        height="16"
        viewBox="0 0 16 16"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M3.33301 8H12.6663"
          stroke="white"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </button>
    <div
      class="AmountCounterInputWrapper"
    >
      <input
        class="AmountCounterInput"
        data-testid="AmountCounter.input"
        min="1"
        readonly=""
        type="number"
        value="1"
      />
    </div>
    <button
      class="AmountCounterButton"
      data-testid="AmountCounter.plusButton"
    >
      <svg
        class="AmountCounterButtonIcon"
        fill="none"
        viewBox="0 0 14 14"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M2.33398 6.99998H11.6673M7.00065 2.33331V11.6666V2.33331Z"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="1.5"
        />
      </svg>
    </button>
  </div>
</div>
`;
